File: NeetCode/Swift/Stack/Medium/155.MinStack.py
# Design a stack that supports push, pop, top, 
# and retrieving the minimum element in constant time.

class MinStack(object):
    def __init__(self):
        self.stack = []
        self.minimum = []

    def push(self, val):
        self.stack.append(val)

        if self.minimum and val > self.minimum[-1]:
            self.minimum.append(self.minimum[-1])
        else:
            self.minimum.append(val)
        
    def pop(self):
        self.stack.pop()
        self.minimum.pop()

    def top(self):
        return self.stack[-1]

    def getMin(self):
        return self.minimum[-1]

File: swift/Neetcode/Stack/Medium/155.MinStack.swift
// Design a stack that supports push, pop, top, 
// and retrieving the minimum element in constant time.

class MinStack {
    
    private var stack: [(Int, Int)]

    init() {
        self.stack = []
    }
    
    func push(_ val: Int) {
        if let last = stack.last, val > last.1 {
            stack.append((val, last.1))
        } else {
            stack.append((val, val))
        }
    }
    
    func pop() {
        if !stack.isEmpty { stack.removeLast() }
    }
    
    func top() -> Int {
        return stack.last?.0 ?? 0
    }
    
    func getMin() -> Int {
        return stack.last?.1 ?? 0
    }
}